File: legacy/src/main/rs/emulate/legacy/version/MapIndexEncoder.java
package rs.emulate.legacy.version;

import java.util.Arrays;

import rs.emulate.shared.util.DataBuffer;

/**
 * An encoder for a {@link MapIndex}.
 *
 * @author Major
 */
public final class MapIndexEncoder {

	/**
	 * The MapIndex to encode.
	 */
	private final MapIndex index;

	/**
	 * Creates the MapIndexEncoder.
	 *
	 * @param index The {@link MapIndex} to encode.
	 */
	public MapIndexEncoder(MapIndex index) {
		this.index = index;
	}

	/**
	 * Encodes the {@link MapIndex} into a {@link DataBuffer}.
	 * 
	 * @return The DataBuffer.
	 */
	public DataBuffer encode() {
		DataBuffer buffer = DataBuffer.allocate((3 * Short.BYTES + Byte.BYTES) * index.getSize());

		Arrays.stream(index.getAreas()).forEach(buffer::putShort);
		Arrays.stream(index.getMaps()).forEach(buffer::putShort);
		Arrays.stream(index.getObjects()).forEach(buffer::putShort);

		for (boolean members : index.getMembers()) {
			buffer.putBoolean(members);
		}

		return buffer.flip();
	}

}

File: legacy/src/main/kotlin/rs/emulate/legacy/version/map/MapIndexEncoder.kt
package rs.emulate.legacy.version.map

import rs.emulate.util.putBoolean
import java.nio.ByteBuffer

/**
 * An encoder for a [MapIndex].
 */
class MapIndexEncoder(private val index: MapIndex) {

    /**
     * Encodes the [MapIndex] into a [ByteBuffer].
     */
    fun encode(): ByteBuffer {
        val buffer = ByteBuffer.allocate((3 * java.lang.Short.BYTES + java.lang.Byte.BYTES) * index.size)

        index.areas.forEach { buffer.putShort(it.toShort()) }
        index.maps.forEach { buffer.putShort(it.toShort()) }
        index.objects.forEach { buffer.putShort(it.toShort()) }
        index.members.forEach { buffer.putBoolean(it) }

        return buffer.apply { flip() }
    }

}
